### PR TITLE
Feat: create grpc binding functions for identity service

### PR DESCRIPTION
**Pull request by Nhsdkk**
1) Create binding functions
2) Make permission restrictions
3) Move from redis to postgres for storing refresh tokens
4) Add option to get all user
5) Make all fields modifiable for user

### CODE REVIEW

File: identity_service/repository/mock_token_repository.py [hunk 1]
```diff
@@ -61,21 +74,56 @@ async def get_refresh_token(self, user_id: str) -> str:
             No refresh token found for provided user_id
 
         """
-        result = self._tokens.get(user_id)
+        result = None
+
+        for user in self._tokens.values():
```
Comment (ZixMai) [inline identity_service/repository/mock_token_repository.py]: maybe use next(user[session_id] for user in self._tokens.values() if session_id in user) and try except StopIteration instead

File: identity_service/repository/mock_user_repository.py [hunk 2]
```diff
@@ -140,12 +164,15 @@ async def update_user(self, user: User) -> None:
...
         ------
-        ValueError
+        ValueNotFoundError
             Can't update user with provided data
 
         """
-        index: int = self._users.index(user)
-        self._users[index] = user
+        try:
+            index: int = self._users.index(user)
```
Comment (ZixMai) [inline identity_service/repository/mock_user_repository.py:185]: wrong way to find index, check pr where ive pinged you

File: identity_service/repository/mock_user_repository.py [hunk 3]
```diff
@@ -76,11 +89,14 @@ async def get_user_by_id(self, user_id: str) -> User:
...
-        ValueError
+        ValueNotFoundError
             User does not exist
 
         """
-        return next(user for user in self._users if user.id == user_id)
+        values = [user for user in self._users if user.id == user_id]
+        if len(values) == 0:
+            raise ValueNotFoundError("No user found for this id")
+        return values[0]
```
Comment (ZixMai) [inline identity_service/repository/mock_user_repository.py]: use next instead list()[0]

File: identity_service/repository/mock_user_repository.py [hunk 4]
```diff
@@ -54,11 +64,14 @@ async def get_user_by_email(self, email: str) -> User:
...
-        ValueError
+        ValueNotFoundError
             User does not exist
 
         """
-        return next(user for user in self._users if user.email == email)
+        values = [user for user in self._users if user.email == email]
+        if len(values) == 0:
+            raise ValueNotFoundError("No user found for this id")
+        return values[0]
```
Comment (ZixMai) [inline identity_service/repository/mock_user_repository.py]: use next() instead list()[0]

File: identity_service/repository/token_repository_impl.py [hunk 5]
```diff
@@ -26,40 +24,48 @@ class TokenRepositoryImpl:
         Stores refresh token in Redis database
     async delete_refresh_token(user_id)
         Deletes refresh token corresponding to provided user_id
-
+    async delete_all_refresh_tokens(user_id)
+        Delete all user's refresh tokens
     """
```
Comment (ZixMai) [inline identity_service/repository/token_repository_impl.py:30]: add empty line

File: identity_service/utils/jwt_controller.py [hunk 6]
```diff
@@ -46,40 +46,47 @@ def __init__(self) -> None:
         self._access_key = os.environ["ACCESS_SECRET"]
         self._refresh_key = os.environ["REFRESH_SECRET"]
 
-    def generate(self, user_id: str) -> Tuple[str, str]:
+    def generate_access_token(self, user_id: str, session_id: str) -> str:
+        access_token: str = encode(
```
Comment (ZixMai) [inline identity_service/utils/jwt_controller.py:68]: add docstring?

**Review by ZixMai — CHANGES_REQUESTED** [review]

File: identity_service/repository/mock_token_repository.py (hunk 1 above)
Comment (Nhsdkk) [reply identity_service/repository/mock_token_repository.py]: fixed

File: identity_service/repository/mock_user_repository.py (hunk 2 above)
Comment (Nhsdkk) [reply identity_service/repository/mock_user_repository.py:185]: wont fix

File: identity_service/repository/mock_user_repository.py (hunk 3 above)
Comment (Nhsdkk) [reply identity_service/repository/mock_user_repository.py]: fixed

File: identity_service/repository/mock_user_repository.py (hunk 4 above)
Comment (Nhsdkk) [reply identity_service/repository/mock_user_repository.py]: fixed

File: identity_service/repository/token_repository_impl.py (hunk 5 above)
Comment (Nhsdkk) [reply identity_service/repository/token_repository_impl.py:30]: fixed

File: identity_service/utils/jwt_controller.py (hunk 6 above)
Comment (Nhsdkk) [reply identity_service/utils/jwt_controller.py:68]: fixed

File: identity_service/repository/mock_token_repository.py [hunk 7]
```diff
@@ -1,54 +1,67 @@
...
+        Stores refresh token
+    async delete_refresh_token(session_id)
+        Deletes refresh token corresponding to provided session_id
+    async delete_all_refresh_tokens(user_id)
+        Deletes all user's refresh tokens
 
     """
 
-    _tokens: dict[str, str]
+    _tokens: dict[str, dict[str, str]]
```
Comment (MaxPlays35) [inline identity_service/repository/mock_token_repository.py:31]: We use python generics or from typing?

File: identity_service/repository/mock_token_repository.py [hunk 8]
```diff
@@ -1,54 +1,67 @@
...
+@singleton
+class MockTokenRepositoryImpl(TokenRepositoryInterface):
     """
     Mock class for manipulating with token data
 
     Attributes
     ----------
-    _tokens: dict[str, str]
-        Dictionary with pairs of [user_id, refresh_token]
+    _tokens: dict[str, dict[str, str]]
```
Comment (MaxPlays35) [inline identity_service/repository/mock_token_repository.py:15]: Same here

File: identity_service/repository/mock_token_repository.py (hunk 7 above)
Comment (Nhsdkk) [reply identity_service/repository/mock_token_repository.py:31]: already in issues

File: identity_service/repository/mock_token_repository.py (hunk 8 above)
Comment (Nhsdkk) [reply identity_service/repository/mock_token_repository.py:15]: already in issues

**Review by MaxPlays35 — APPROVED** [review]

File: identity_service/src/identity_service_impl.py [hunk 9]
```diff
@@ -51,16 +69,18 @@ def __init__(
...
+    async def login(
+        self, request: auth_proto.LoginRequest, context: grpc.ServicerContext
     ) -> auth_proto.CredentialsResponse:
         """
         Log the user in and return credentials if user data matches
 
         Parameters
         ----------
-        _ : LoginRequest
+        request : LoginRequest
```
Comment (ZixMai) [inline identity_service/src/identity_service_impl.py]: wrong type of request and return in docstring

File: identity_service/src/identity_service_impl.py [hunk 10]
```diff
@@ -71,13 +91,41 @@ def login(
...
+        except prisma.errors.PrismaError:
+            context.set_code(grpc.StatusCode.INTERNAL)
+            return auth_proto.CredentialsResponse(
+                status_code=500, message="Internal server error"
+            )
+
+    async def register(
         self, request: auth_proto.RegisterRequest, context: grpc.ServicerContext
     ) -> auth_proto.CredentialsResponse:
         """
```
Comment (ZixMai) [inline identity_service/src/identity_service_impl.py:131]: wrong type of request and return in docstring

File: identity_service/src/identity_service_impl.py [hunk 11]
```diff
@@ -96,11 +144,41 @@ def register(
...
+            return auth_proto.CredentialsResponse(
+                status_code=500, message="Internal server error"
+            )
+
+    async def auth(
         self, request: auth_proto.AccessToken, context: grpc.ServicerContext
-    ) -> auth_proto.AuthResponse:
+    ) -> get_user_proto.UserResponse:
         """
         Authenticates user by his token and returns his ID
```
Comment (ZixMai) [inline identity_service/src/identity_service_impl.py:184]: wrong type of request and return in docstring

File: identity_service/src/identity_service_impl.py [hunk 12]
```diff
@@ -140,11 +237,39 @@ def get_new_access_token(
             Response object with new access token
```
Comment (ZixMai) [inline identity_service/src/identity_service_impl.py:237]: wrong type of request and return in docstring

File: identity_service/src/identity_service_impl.py [hunk 13]
```diff
@@ -140,11 +237,39 @@ def get_new_access_token(
...
+            return get_access_token_proto.GetNewAccessTokenResponse(
+                status_code=500, message="Internal server error"
+            )
+
+    async def get_user_by_id(
         self, request: get_user_proto.UserByIdRequest, context: grpc.ServicerContext
-    ) -> get_user_proto.UserByIdResponse:
+    ) -> get_user_proto.UserResponse:
         """
         Gets user object that matches given ID
```
Comment (ZixMai) [inline identity_service/src/identity_service_impl.py:273]: wrong type of request and return in docstring

File: identity_service/src/identity_service_impl.py [hunk 14]
```diff
@@ -157,15 +282,30 @@ def get_user_by_id(
...
+            return get_user_proto.UserResponse(
+                status_code=500, message="Internal server error"
+            )
+
+    async def get_users_by_id(
         self, request: get_user_proto.UsersByIdRequest, context: grpc.ServicerContext
-    ) -> get_user_proto.UsersByIdResponse:
+    ) -> get_user_proto.UsersResponse:
         """
         Gets user objects that matches given ids
```
Comment (ZixMai) [inline identity_service/src/identity_service_impl.py:309]: wrong type of request and return in docstring

File: identity_service/src/identity_service_impl.py [hunk 15]
```diff
@@ -178,24 +318,87 @@ def get_users_by_id(
...
+        Parameters
+        ----------
+        request : get_user_proto.GetAllUsersRequest
+            Requesting user's data
+        context : grpc.ServicerContext
+            Request context
+
+        Returns
+        -------
+        UsersResponse
```
Comment (ZixMai) [inline identity_service/src/identity_service_impl.py]: wrong type of return in docstring

File: identity_service/src/identity_service_impl.py [hunk 16]
```diff
@@ -178,24 +318,87 @@ def get_users_by_id(
...
         request: update_user_proto.UpdateUserRequest,
         context: grpc.ServicerContext,
-    ) -> requests_proto.BaseResponse:
+    ) -> auth_proto.CredentialsResponse:
         """
         Updates user data
 
         Parameters
         ----------
         request : UpdateUserRequest
```
Comment (ZixMai) [inline identity_service/src/identity_service_impl.py]: wrong type of request and return in docstring

File: identity_service/src/identity_service_impl.py [hunk 17]
```diff
@@ -218,7 +472,7 @@ def delete_user(
         Parameters
         ----------
         request : DeleteUserRequest
```
Comment (ZixMai) [inline identity_service/src/identity_service_impl.py]: wrong type of request and return in docstring

File: identity_service/src/identity_service_impl.py [hunk 18]
```diff
@@ -228,9 +482,32 @@ def delete_user(
...
+        except prisma.errors.PrismaError:
+            context.set_code(grpc.StatusCode.INTERNAL)
+            return requests_proto.BaseResponse(
+                status_code=500, message="Internal server error"
+            )
+
+    async def logout(
         self, request: auth_proto.AccessToken, context: grpc.ServicerContext
     ) -> requests_proto.BaseResponse:
         """
```
Comment (ZixMai) [inline identity_service/src/identity_service_impl.py:511]: wrong type of request and return in docstring

**Review by ZixMai — CHANGES_REQUESTED** [review]
:^D

**Review by ZixMai — APPROVED** [review]
GG

**Review by ZixMai — APPROVED** [review]

**Pull request merged**